### PR TITLE
Adding UC thermal properties in solid_properties module

### LINKED ISSUE #23207: Add Material Properties for Uranium Monocarbide
## Reason
Adding the thermal material properties (temperature dependent) of Uranium Monocarbide in the solid_properties module.  
The Userobjects available in the Solid Properties module that provide thermal properties are:
[graphite](https://mooseframework.inl.gov/source/userobjects/ThermalGraphiteProperties.html)
[monolithic silicon carbide](https://mooseframework.inl.gov/source/userobjects/ThermalMonolithicSiCProperties.html)
[composite silicon carbide](https://mooseframework.inl.gov/source/userobjects/ThermalCompositeSiCProperties.html)
[stainless steel alloy 316](https://mooseframework.inl.gov/source/userobjects/ThermalSS316Properties.html)
[using functions](https://mooseframework.inl.gov/source/userobjects/ThermalFunctionSolidProperties.html)

## Design
A PR was implemented to add this material.
EDIT: a new solid properties user object
Adding more Userobjects in the Solid Properties module.

## Impact
EDIT:
Provides the properties to all downstream apps for nuclear thermal solid simulations

### PR DESCRIPTION
**Pull request by AyaHegazy22**
close #23207

### CODE REVIEW

**Comment by AyaHegazy22** [comment]
@dschwen Thank you so much for your help and comments, I created another PR to fix the problem with the other PR. I tried the commands you posted in the comments but I still wasn't not able to fix the problem. I lack the experience with git and I tried different things, so please let me know if it still has a problem, thanks again for your great help, I greatly appreciate it.

**Comment by dschwen** [comment]
Yeah, that looks much better. One thing you need to fix is putting the issue number in at least one of the commit messages. Our SQA rules require that.

**Comment by GiudGiud** [comment]
`git commit --amend` to modify the last commit message.
Then add : "refs #23207" to the message

File: modules/solid_properties/include/userobjects/ThermalUCProperties.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+class ThermalUCProperties : public ThermalSolidProperties
+{
+public:
+  static InputParameters validParams();
+
+  ThermalUCProperties(const InputParameters & parameters);
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Woverloaded-virtual"
```
Comment (dschwen) [inline modules/solid_properties/include/userobjects/ThermalUCProperties.h:13]: Why is this required?

Comment (dschwen) [reply modules/solid_properties/include/userobjects/ThermalUCProperties.h:13]: Short answer: it is not. If the overload is intentional a `using Base::member` declaration should be added.

**Comment by aprilnovak** [comment]
@GiudGiud or @dschwen can someone please add Aya as a contributor so the test suite will launch?

Thanks!

**Comment by GiudGiud** [comment]
@loganharbour has rights to do that
@socratesgorilla can you as well?

**Comment by aprilnovak** [comment]
Oh I might be misinterpreting the failed tests - @AyaHegazy22 I think you need to do another clang format. Perhaps that's why the tests aren't launching.

File: modules/solid_properties/src/userobjects/ThermalUCProperties.C [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+registerMooseObject("SolidPropertiesApp", ThermalUCProperties);
+
+InputParameters
+ThermalUCProperties::validParams()
+
+{
+  InputParameters params = ThermalSolidProperties::validParams();
+
+  params.addRangeCheckedParam<Real>("density", 13824.7, "density > 0.0", "(Constant) density");
```
Comment (GiudGiud) [inline modules/solid_properties/src/userobjects/ThermalUCProperties.C:12]: add unit

File: modules/solid_properties/include/userobjects/ThermalUCProperties.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
Comment (GiudGiud) [inline modules/solid_properties/include/userobjects/ThermalUCProperties.h:1]: add moose copyright header

same for other files

File: modules/solid_properties/doc/content/source/userobjects/ThermalUCProperties.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+## Description
+
+This userobject provides
+thermal properties for Uranium monocarbide as a function of temperature.
+
+!include solid_properties_units.md
+
+Isobaric specific heat is calculated from [!cite](iaea) as
+
+\begin{equation}
```
Comment (GiudGiud) [inline modules/solid_properties/doc/content/source/userobjects/ThermalUCProperties.md:14]: I m not sure that s the syntax, we re getting a latex build error
maybe !equation ? what do the other pages have

**Review by GiudGiud — COMMENTED** [review]
some test failure:
``
/data/civet2/build/moose/modules/solid_properties/unit/src/ThermalUCPropertiesTest.C:13: Failure
Expected: (std::abs(((_sp1->k_from_T(T)) - (21.57840004)) / (21.57840004))) <= (1e-12), actual: 1.85371e-09 vs 1e-12
``

This is simply because the saved value does not have enough digits

File: modules/solid_properties/src/userobjects/ThermalUCProperties.C [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+ThermalUCProperties::ThermalUCProperties(const InputParameters & parameters)
+  : ThermalSolidProperties(parameters), _rho_const(getParam<Real>("density"))
+{
+}
+
+Real
+ThermalUCProperties::cp_from_T(const Real & T) const
+{
+  return 239.7 - 5.068e-3 * T + 1.7604e-5 * Utility::pow<2>(T) - 3488100 / Utility::pow<2>(T);
```
Comment (GiudGiud) [inline modules/solid_properties/src/userobjects/ThermalUCProperties.C:25]: add an if on range of validity
```
298 K $\le$ T $\le$ 2838 K
```
and declare solution is invalid outside. 

You may need to add the SOlutionInvalidInterface first to the classes that the SolidFluidProperties first
Please let me know if you need help with that

**Comment by GiudGiud** [comment]
@aprilnovak what is the status here? Should we take this over or did it get added elsewhere?

File: modules/solid_properties/src/userobjects/ThermalUCProperties.C (hunk 5 above)
Comment (GiudGiud) [reply modules/solid_properties/src/userobjects/ThermalUCProperties.C:25]: https://github.com/idaholab/moose/blob/next/modules/fluid_properties/src/fluidproperties/NaKFluidProperties.C
for an example of doing this

File: modules/solid_properties/include/userobjects/ThermalUCProperties.h (hunk 3 above)
Comment (GiudGiud) [reply modules/solid_properties/include/userobjects/ThermalUCProperties.h:1]: also solid properties are now in the solidproperties folder no longer in userobjects/

use `git mv` to move them pls

**Comment by tophmatthews** [comment]
Note, some of these things are in BISON already, which is the appropriate place for such things, right?

**Comment by tophmatthews** [comment]
https://mooseframework.inl.gov/bison/source/materials/MCThermal.html#mcthermal

**Comment by aprilnovak** [comment]
We would like to use these properties for simulations not involving BISON (such as with the heat conduction module).

**Comment by GiudGiud** [comment]
If a new or existing material property can be released and is of interest to the community then it should be. There's no use handicapping our open-source capabilities when we don't need to.

**Comment by aprilnovak** [comment]
Would someone at INL please add Aya to the list of github users for CIVET? I am not sure why this reverted, the test suite did launch properly a few months back.

**Comment by aprilnovak** [comment]
Would someone please add Aya to the CIVET list? I believe this is ready to go.

**Review by GiudGiud — APPROVED** [review]
looks great. sorry fell off my radar

**Review by GiudGiud — APPROVED** [review]
is it me or the IAEA reference is wrong for the unit of Cp ?
Eq. 2.81 of https://www-pub.iaea.org/MTCD/Publications/PDF/IAEA-THPH_web.pdf
shows it per g I think

**Comment by aprilnovak** [comment]
@GiudGiud I agree, it looks like the IAEA doc has a typo, 0.22 J/kg/K would be really low. It looks like the actual code in the user object assumed that the IAEA doc has a typo - @AyaHegazy22 is that correct?

**Comment by AyaHegazy22** [comment]
Yeah, that's correct.
